***
<div class="alert alert-block alert-success">
<font color = '#006400'> 
    
# 1. Data Collection </font>    

Importing Libraries, reading Data and merging Datasets.
</div>

In [63]:
# Import Libraries
import pandas as pd
import numpy as np

In [3]:
# Read Data
demographic = pd.read_excel("Telco_customer_churn_demographics.xlsx")
location = pd.read_excel("Telco_customer_churn_location.xlsx")
population = pd.read_excel("Telco_customer_churn_population.xlsx")
services = pd.read_excel("Telco_customer_churn_services.xlsx")
status = pd.read_excel("Telco_customer_churn_status.xlsx")

In [34]:
# Display
pd.set_option('display.max_columns', None)

display(demographic.head(2))
display(location.head(2))
display(population.head(2))
display(services.head(2))
display(status.head(2))

Customer ID  Count  Gender  Age Under 30 Senior Citizen Married Dependents  \
0  8779-QRDMV      1    Male   78       No            Yes      No         No   
1  7495-OOKFY      1  Female   74       No            Yes     Yes        Yes   

   Number of Dependents  
0                     0  
1                     1

Customer ID  Count        Country       State         City  Zip Code  \
0  8779-QRDMV      1  United States  California  Los Angeles     90022   
1  7495-OOKFY      1  United States  California  Los Angeles     90063   

                 Lat Long   Latitude   Longitude  
0   34.02381, -118.156582  34.023810 -118.156582  
1  34.044271, -118.185237  34.044271 -118.185237

ID  Zip Code  Population
0   1     90001       54492
1   2     90002       44586

Customer ID  Count Quarter Referred a Friend  Number of Referrals  \
0  8779-QRDMV      1      Q3                No                    0   
1  7495-OOKFY      1      Q3               Yes                    1   

   Tenure in Months    Offer Phone Service  Avg Monthly Long Distance Charges  \
0                 1     None            No                               0.00   
1                 8  Offer E           Yes                              48.85   

  Multiple Lines Internet Service Internet Type  Avg Monthly GB Download  \
0             No              Yes           DSL                        8   
1            Yes              Yes   Fiber Optic                       17   

  Online Security Online Backup Device Protection Plan Premium Tech Support  \
0              No            No                    Yes                   No   
1              No           Yes                     No                   No   

  Streaming TV Streaming Movies Streaming Music Unlimited Data  \
0           No              Yes              No             No   
1           No               No              No            Yes   

         Contract Paperless Billing   Payment Method  Monthly Charge  \
0  Month-to-Month               Yes  Bank Withdrawal           39.65   
1  Month-to-Month               Yes      Credit Card           80.65   

   Total Charges  Total Refunds  Total Extra Data Charges  \
0          39.65            0.0                        20   
1         633.30            0.0                         0   

   Total Long Distance Charges  Total Revenue  
0                          0.0          59.65  
1                        390.8        1024.10

Customer ID  Count Quarter  Satisfaction Score Customer Status Churn Label  \
0  8779-QRDMV      1      Q3                   3         Churned         Yes   
1  7495-OOKFY      1      Q3                   3         Churned         Yes   

   Churn Value  Churn Score  CLTV Churn Category                  Churn Reason  
0            1           91  5433     Competitor  Competitor offered more data  
1            1           69  5302     Competitor  Competitor made better offer

In [53]:
# Merge the tables on "Customer ID"

df = demographic.merge(location, on='Customer ID', how='outer') \
                       .merge(services, on='Customer ID', how='outer') \
                       .merge(status, on='Customer ID', how='outer')

# Set "Customer ID" as an index

df.set_index('Customer ID', inplace = True)

# Display the first 2 rows of the merged table
display(df.head(2))

C:\Users\saras\AppData\Local\Temp\ipykernel_15968\1014119456.py:3: FutureWarning: Passing 'suffixes' which cause duplicate columns {'Count_x'} in the result is deprecated and will raise a MergeError in a future version.
  df = demographic.merge(location, on='Customer ID', how='outer') \


Count_x  Gender  Age Under 30 Senior Citizen Married Dependents  \
Customer ID                                                                    
8779-QRDMV         1    Male   78       No            Yes      No         No   
7495-OOKFY         1  Female   74       No            Yes     Yes        Yes   

             Number of Dependents  Count_y        Country       State  \
Customer ID                                                             
8779-QRDMV                      0        1  United States  California   
7495-OOKFY                      1        1  United States  California   

                    City  Zip Code                Lat Long   Latitude  \
Customer ID                                                             
8779-QRDMV   Los Angeles     90022   34.02381, -118.156582  34.023810   
7495-OOKFY   Los Angeles     90063  34.044271, -118.185237  34.044271   

              Longitude  Count_x Quarter_x Referred a Friend  \
Customer ID                                                    
8779-QRDMV  -118.156582        1        Q3                No   
7495-OOKFY  -118.185237        1        Q3               Yes   

             Number of Referrals  Tenure in Months    Offer Phone Service  \
Customer ID                                                                 
8779-QRDMV                     0                 1     None            No   
7495-OOKFY                     1                 8  Offer E           Yes   

             Avg Monthly Long Distance Charges Multiple Lines  \
Customer ID                                                     
8779-QRDMV                                0.00             No   
7495-OOKFY                               48.85            Yes   

            Internet Service Internet Type  Avg Monthly GB Download  \
Customer ID                                                           
8779-QRDMV               Yes           DSL                        8   
7495-OOKFY               Yes   Fiber Optic                       17   

            Online Security Online Backup Device Protection Plan  \
Customer ID                                                        
8779-QRDMV               No            No                    Yes   
7495-OOKFY               No           Yes                     No   

            Premium Tech Support Streaming TV Streaming Movies  \
Customer ID                                                      
8779-QRDMV                    No           No              Yes   
7495-OOKFY                    No           No               No   

            Streaming Music Unlimited Data        Contract Paperless Billing  \
Customer ID                                                                    
8779-QRDMV               No             No  Month-to-Month               Yes   
7495-OOKFY               No            Yes  Month-to-Month               Yes   

              Payment Method  Monthly Charge  Total Charges  Total Refunds  \
Customer ID                                                                  
8779-QRDMV   Bank Withdrawal           39.65          39.65            0.0   
7495-OOKFY       Credit Card           80.65         633.30            0.0   

             Total Extra Data Charges  Total Long Distance Charges  \
Customer ID                                                          
8779-QRDMV                         20                          0.0   
7495-OOKFY                          0                        390.8   

             Total Revenue  Count_y Quarter_y  Satisfaction Score  \
Customer ID                                                         
8779-QRDMV           59.65        1        Q3                   3   
7495-OOKFY         1024.10        1        Q3                   3   

            Customer Status Churn Label  Churn Value  Churn Score  CLTV  \
Customer ID                                                               
8779-QRDMV          Churned         Yes            1           91  5433   
7495-OOKFY          Churned         Yes

In [54]:
# Check for duplicates

if df.duplicated().any():
    display(df[df.duplicated(keep=False)])
else:
    print("No duplicates found")

No duplicates found


  
<div class="alert alert-block alert-success">
<a id='explore'>
<font color = '#006400'> 
    
# 2. Data Exploration </font>
</a>

Understand the structure of the data and provide the first impressions of patterns, relations and errors.
    
</div>

### 2.1. Basic info and statistics

In [58]:
# Drop repeated columns "Count" and "Quarter"

df = df.drop(columns=['Count_x', 'Count_y', 'Quarter_x'])
df = df.rename(columns={'Quarter_y': 'Quarter'})

In [59]:
# Checking df general info in a dataframe

def df_info(data): 
    '''
    Returns a dataframe of .info() method, instead of its usual design
    
    Args: 
        data (panda dataframe) - data to analyse
    Returns: 
        pandas dataframe'''
    
    dtype_info = pd.DataFrame(data.dtypes).T.rename(index={0:'Column type'})
    missing_val_info = pd.DataFrame(data.isna().sum()).T.rename(index={0:'Count for missing values'})
    missing_val_percent_info = pd.DataFrame(data.isna().sum() / len(data)*100).T.rename(index={0:'Missing values (in %)'})
    
    tab_info = pd.concat([dtype_info, missing_val_info, missing_val_percent_info], axis=0)
    tab_info = tab_info.T.join(data.describe().T).fillna('-')
    
    return tab_info


display(df_info(df).style)

# Display shape of df:
print(f'\nThe DataFrame has {df.shape[1]} features with data from {df.shape[0]} rows.')


The DataFrame has 50 features with data from 7043 rows.


In [60]:
# For qualitative features

df.describe(include='O').T

count unique                            top  freq
Gender                  7043      2                           Male  3555
Under 30                7043      2                             No  5642
Senior Citizen          7043      2                             No  5901
Married                 7043      2                             No  3641
Dependents              7043      2                             No  5416
Country                 7043      1                  United States  7043
State                   7043      1                     California  7043
City                    7043   1106                    Los Angeles   293
Lat Long                7043   1679         33.362575, -117.299644    43
Referred a Friend       7043      2                             No  3821
Offer                   7043      6                           None  3877
Phone Service           7043      2                            Yes  6361
Multiple Lines          7043      2                             No  4072
Internet Service        7043      2                            Yes  5517
Internet Type           7043      4                    Fiber Optic  3035
Online Security         7043      2                             No  5024
Online Backup           7043      2                             No  4614
Device Protection Plan  7043      2                             No  4621
Premium Tech Support    7043      2                             No  4999
Streaming TV            7043      2                             No  4336
Streaming Movies        7043      2                             No  4311
Streaming Music         7043      2                             No  4555
Unlimited Data          7043      2                            Yes  4745
Contract                7043      3                 Month-to-Month  3610
Paperless Billing       7043      2                            Yes  4171
Payment Method          7043      3                Bank Withdrawal  3909
Quarter                 7043      1                             Q3  7043
Customer Status         7043      3                         Stayed  4720
Churn Label             7043      2                             No  5174
Churn Category          1869      5                     Competitor   841
Churn Reason            1869     20  Competitor had better devices   313

__Notes:__

> - There are __7043 rows__ (Clients) and __50 features__ <p>
> - There are __19 quantitative variables__ and __31 qualitative__ <p>
> - __Missing values:__ Churn Category (5174) and Churn reason (5174)  <p>
> - __Constant variables:__ Country and State have the unique values "United States" and "California" respectively; Quarter is always Q3 <p>
> - City has 1106 different values. But California only has 482 cities <p>
> - __Check if categoric values are acceptable:__ Offer has 6 possible values; Contract has 3 possivlue values; Payment Method has 3 possible values; Customer Status has 3 possible values; Churn Category has 5 possible values <p>
> - For age information we have two dummy variables: Under 30 and Senior Citizen <p>
> - Check if when Dependents is Yes then Number of Dependents is higher than 0 <p>

### 2.2. Further exploration 

__Skewness and Kurtosis__

> - __Skewness:__ Degree of asymmetry. Between -0.5 and 0.5 then is approx symmetric; 0 is symetric; between -1 and -0.5 or 0.5 and 1 is moderately skewed. Less than -1 or greater than 1 is highly skewed. <p>
> - __Kurtosis:__ Degree of peakedness. High kurtosis (>3) indicate a peaked distribution which could indicate the presence of outliers. <p>


__Notes:__
> - __Highly skewed:__ Number of Dependents, Number of Referrals, Avg Monthly GB Download, Total Refunds, Total Extra Data Charges, Total Long Distance Charges, Churv Value
> - __High kurtosis:__ Total Refunds, Number of Dependents, Total Extra Data Charges

In [64]:
# Compute skewness and kurtosis on numeric variables only

num_cols = df.select_dtypes(include=np.number).columns
tab_info4 = pd.DataFrame(df[num_cols].skew(), columns=['Skewness'])
tab_info5 = pd.DataFrame(df[num_cols].kurtosis(), columns=['Kurtosis'])
tab_info2 = pd.concat([tab_info4,tab_info5])
tab_info2

Skewness   Kurtosis
Age                                0.162186        NaN
Number of Dependents               2.109932        NaN
Zip Code                          -0.209615        NaN
Latitude                           0.314804        NaN
Longitude                         -0.091932        NaN
Number of Referrals                1.446060        NaN
Tenure in Months                   0.240543        NaN
Avg Monthly Long Distance Charges  0.049176        NaN
Avg Monthly GB Download            1.216584        NaN
Monthly Charge                    -0.220524        NaN
Total Charges                      0.963791        NaN
Total Refunds                      4.328517        NaN
Total Extra Data Charges           4.091209        NaN
Total Long Distance Charges        1.238282        NaN
Total Revenue                      0.919410        NaN
Satisfaction Score                -0.367401        NaN
Churn Value                        1.063031        NaN
Churn Score                       -0.156852        NaN
CLTV                              -0.311602        NaN
Age                                     NaN  -1.002849
Number of Dependents                    NaN   4.446358
Zip Code                                NaN  -1.173915
Latitude                                NaN  -1.160506
Longitude                               NaN  -1.191291
Number of Referrals                     NaN   0.721964
Tenure in Months                        NaN  -1.387052
Avg Monthly Long Distance Charges       NaN  -1.254654
Avg Monthly GB Download                 NaN   0.881502
Monthly Charge                          NaN  -1.257260
Total Charges                           NaN  -0.227693
Total Refunds                           NaN  18.350658
Total Extra Data Charges                NaN  16.458874
Total Long Distance Charges             NaN   0.644092
Total Revenue                           NaN  -0.203457
Satisfaction Score                      NaN  -0.536853
Churn Value                             NaN  -0.870211
Churn Score                             NaN  -1.085290
CLTV                                    NaN  -0.934032

__Categorical Variables - Check levels/possible values__

Offer has 6 possible values; Contract has 3 possivlue values; Payment Method has 3 possible values; Customer Status has 3 possible values; Churn Category has 5 possible values

In [67]:
df['Offer'].value_counts()

None       3877
Offer B     824
Offer E     805
Offer D     602
Offer A     520
Offer C     415
Name: Offer, dtype: int64

In [72]:
merged_df['Churn Reason'].value_counts()

Churn Reason
Competitor had better devices                313
Competitor made better offer                 311
Attitude of support person                   220
Don't know                                   130
Competitor offered more data                 117
Competitor offered higher download speeds    100
Attitude of service provider                  94
Price too high                                78
Product dissatisfaction                       77
Network reliability                           72
Long distance charges                         64
Service dissatisfaction                       63
Moved                                         46
Extra data charges                            39
Limited range of services                     37
Poor expertise of online support              31
Lack of affordable download/upload speed      30
Lack of self-service on Website               29
Poor expertise of phone support               12
Deceased                                       6
Name: c

In [73]:
merged_df['Payment Method'].value_counts()

Payment Method
Bank Withdrawal    3909
Credit Card        2749
Mailed Check        385
Name: count, dtype: int64

<div class="alert alert-block alert-warning">

<a id='visual'></a>

## 2.3. Visual Exploration
    
</div>    

<div class="alert alert-block alert-info">
    
<a id='visual_num'></a>

### 2.3.1. Numerical Variables
    
</div>
